api/prometheus.go

```diff
@@ -0,0 +1,453 @@
+			case promql.ErrStorage:
+				response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+					Status:    statusError,
+					Error:     fmt.Errorf("query failed: %v", res.Err),
+					ErrorType: errorInternal,
+				}, ""))
+			}
+			response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+				Status:    statusError,
+				Error:     fmt.Errorf("query failed: %v", res.Err),
+				ErrorType: errorExec,
+			}, ""))
+		}
+		return
+	}
+
+	response.Write(ctx, response.NewJson(200,
+		prometheusQueryResult{
+			Data: prometheusQueryData{
+				ResultType: res.Value.Type(),
+				Result:     res.Value,
+			},
+			Status: statusSuccess,
+		},
+		"",
+	))
+}
+
+func (s *Server) prometheusQuerySeries(ctx *middleware.Context, request models.PrometheusSeriesQuery) {
+	start, err := parseTime(request.Start)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, fmt.Sprintf("could not parse start time: %v", err)))
+		return
+	}
+
+	end, err := parseTime(request.End)
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusBadRequest, fmt.Sprintf("could not parse end time: %v", err)))
+		return
+	}
+
+	newCtx := context.WithValue(ctx.Req.Context(), orgID("org-id"), ctx.OrgId)
+
+	q := querier{
+		*s,
+		uint32(start.Unix()),
+		uint32(end.Unix()),
+		ctx.OrgId,
+		true,
+		newCtx,
+	}
+	if err != nil {
+		response.Write(ctx, response.NewError(http.StatusInternalServerError, fmt.Sprintf("query failed: %v", err)))
+		return
+	}
+	var matcherSets [][]*labels.Matcher
+	for _, s := range request.Match {
+		matchers, err := promql.ParseMetricSelector(s)
+		if err != nil {
+			response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+				Status:    statusError,
+				Error:     fmt.Errorf("query failed: %v", err),
+				ErrorType: errorBadData,
+			}, ""))
+			return
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
+
+	var sets []storage.SeriesSet
+	for _, mset := range matcherSets {
+		s, err := q.Select(mset...)
+		if err != nil {
+			response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+				Status:    statusError,
+				Error:     fmt.Errorf("query failed: %v", err),
+				ErrorType: errorExec,
+			}, ""))
+			return
+		}
+		sets = append(sets, s)
+	}
+
+	set := storage.NewMergeSeriesSet(sets)
+	metrics := []labels.Labels{}
+	for set.Next() {
+		metrics = append(metrics, set.At().Labels())
+	}
+	if set.Err() != nil {
+		response.Write(ctx, response.NewJson(http.StatusInternalServerError, prometheusQueryResult{
+			Status:    statusError,
+			Error:     fmt.Errorf("query failed: %v", err),
+			ErrorType: errorExec,
+		}, ""))
+		return
+	}
+
+	response.Write(ctx, response.NewJson(200,
+		prometheusQueryResult{
+			Data:   metrics,
+			Status: statusSuccess,
+		},
+		"",
+	))
+
+	return
+}
+
+func (s *Server) Querier(ctx context.Context, mint, maxt int64) (storage.Querier, error) {
```

Comment:
this function should be called NewQuerier I think, and should be under the querier type

Reply:
i see now this won't work, as promql dictates the Querier method name